From 80b8714606aa055123102e50e6be893ee2f652dd Mon Sep 17 00:00:00 2001
From: Matthew Schwartz <matthew.schwartz@linux.dev>
Date: Sat, 31 Jan 2026 16:32:13 -0800
Subject: [PATCH 075/188] mmc: rtsx: add quirk table to disable runtime power
 management
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add a quirk table to disable runtime power management for handheld devices
that we know have Realtek SD card readers.

Link: https://lore.kernel.org/all/ff9a7c20-f465-4afa-bf29-708d4a52974a@linux.dev/
Signed-off-by: Matthew Schwartz <matthew.schwartz@linux.dev>
(cherry picked from commit 96749fad295ae80029ecb0d9f5a53e6620b4229b)
Signed-off-by: Nícolas F. R. A. Prado <nfraprado@collabora.com>
---
 drivers/mmc/host/rtsx_pci_sdmmc.c | 108 +++++++++++++++++++++++++++++-
 1 file changed, 107 insertions(+), 1 deletion(-)

diff --git a/drivers/mmc/host/rtsx_pci_sdmmc.c b/drivers/mmc/host/rtsx_pci_sdmmc.c
index 8dfbc62f1..1fae05e6f 100644
--- a/drivers/mmc/host/rtsx_pci_sdmmc.c
+++ b/drivers/mmc/host/rtsx_pci_sdmmc.c
@@ -7,6 +7,7 @@
  *   Wei WANG <wei_wang@realsil.com.cn>
  */
 
+#include <linux/dmi.h>
 #include <linux/pci.h>
 #include <linux/module.h>
 #include <linux/slab.h>
@@ -23,6 +24,105 @@
 #include <linux/unaligned.h>
 #include <linux/pm_runtime.h>
 
+#define QUIRK_NO_AGGRESSIVE_PM	BIT(0)
+
+/* DMI quirk table for systems with known SD card reader issues. */
+static const struct dmi_system_id rtsx_pci_sdmmc_dmi_quirks[] = {
+	{
+		/* ASUS ROG Xbox Ally */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "ASUSTeK COMPUTER INC."),
+			DMI_MATCH(DMI_BOARD_NAME, "RC73YA"),
+		},
+	},
+	{
+		/* ASUS ROG Xbox Ally X */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "ASUSTeK COMPUTER INC."),
+			DMI_MATCH(DMI_BOARD_NAME, "RC73XA"),
+		},
+	},
+	{
+		/* Lenovo Legion Go */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "83E1"),
+		},
+	},
+	{
+		/* Lenovo Legion Go 2 */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "83N0"),
+		},
+	},
+	{
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "83N1"),
+		},
+	},
+	{
+		/* Lenovo Legion Go S */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "83L3"),
+		},
+	},
+	{
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "83N6"),
+		},
+	},
+	{
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "83Q2"),
+		},
+	},
+	{
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "LENOVO"),
+			DMI_MATCH(DMI_PRODUCT_NAME, "83Q3"),
+		},
+	},
+	{
+		/* MSI Claw */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "Micro-Star International Co., Ltd."),
+			DMI_MATCH(DMI_BOARD_NAME, "MS-1T41"),
+		},
+	},
+	{
+		/* MSI Claw 7 AI+ A2VM */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "Micro-Star International Co., Ltd."),
+			DMI_MATCH(DMI_BOARD_NAME, "MS-1T42"),
+		},
+	},
+	{
+		/* MSI Claw 8 AI+ A2VM */
+		.driver_data = (void *)QUIRK_NO_AGGRESSIVE_PM,
+		.matches = {
+			DMI_MATCH(DMI_BOARD_VENDOR, "Micro-Star International Co., Ltd."),
+			DMI_MATCH(DMI_BOARD_NAME, "MS-1T52"),
+		},
+	},
+	{},
+};
+
 struct realtek_pci_sdmmc {
 	struct platform_device	*pdev;
 	struct rtsx_pcr		*pcr;
@@ -1414,6 +1514,12 @@ static void realtek_init_host(struct realtek_pci_sdmmc *host)
 {
 	struct mmc_host *mmc = host->mmc;
 	struct rtsx_pcr *pcr = host->pcr;
+	const struct dmi_system_id *dmi_id;
+	unsigned long quirks = 0;
+
+	dmi_id = dmi_first_match(rtsx_pci_sdmmc_dmi_quirks);
+	if (dmi_id)
+		quirks = (unsigned long)dmi_id->driver_data;
 
 	mmc->f_min = 250000;
 	mmc->f_max = 208000000;
@@ -1421,7 +1527,7 @@ static void realtek_init_host(struct realtek_pci_sdmmc *host)
 	mmc->caps = MMC_CAP_4_BIT_DATA | MMC_CAP_SD_HIGHSPEED |
 		MMC_CAP_MMC_HIGHSPEED | MMC_CAP_BUS_WIDTH_TEST |
 		MMC_CAP_UHS_SDR12 | MMC_CAP_UHS_SDR25;
-	if (pcr->rtd3_en)
+	if (pcr->rtd3_en && !(quirks & QUIRK_NO_AGGRESSIVE_PM))
 		mmc->caps = mmc->caps | MMC_CAP_AGGRESSIVE_PM;
 	mmc->caps2 = MMC_CAP2_NO_PRESCAN_POWERUP | MMC_CAP2_FULL_PWR_CYCLE |
 		MMC_CAP2_NO_SDIO;
-- 
2.55.0

